Temporal: reflect changes to fix tc39/proposal-temporal#1315

guijemont opened the pull request:
This is meant to be applied after tc39/proposal-temporal#2484 is merged.

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/add/calendar-dateadd-called-with-plaindate-instance.js:
```diff
@@ -0,0 +1,18 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal-adddurationtoorsubtractdurationfromplainyearmonth
```
```suggestion
esid: sec-temporal.plainyearmonth.prototype.add
```
This should refer to the section of the API entry point. That way, if AddDurationToOr... gets renamed, the link won't dangle.

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/add/custom-daysInMonth-irrelevant.js:
```diff
@@ -0,0 +1,27 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal-adddurationtoorsubtractdurationfromplainyearmonth
```
```suggestion
esid: sec-temporal.plainyearmonth.prototype.add
```

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/subtract/calendar-dateadd-called-with-plaindate-instance.js:
```diff
@@ -0,0 +1,18 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal-adddurationtoorsubtractdurationfromplainyearmonth
```
```suggestion
esid: sec-temporal.plainyearmonth.prototype.subtract
```

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/subtract/custom-daysInMonth-irrelevant.js:
```diff
@@ -0,0 +1,27 @@
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-temporal-adddurationtoorsubtractdurationfromplainyearmonth
```
```suggestion
esid: sec-temporal.plainyearmonth.prototype.subtract
```

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/add/calendar-datefromfields-called.js:
```diff
@@ -77,7 +77,6 @@ class CustomCalendar extends Temporal.Calendar {
     return new Temporal.PlainMonthDay(isoMonth, isoDay, this, isoYear);
   }
   dateAdd(date, duration, options) {
-    if (duration.months) throw new Error("adding months not implemented in this test");
```
We would probably want to implement this properly rather than mixing ISO 8601 dateAdd with the weird month scheme in this test calendar.

Off the top of my head, you can probably do it something like this:
```js
const {isoYear, isoMonth, isoDay} = date.getISOFields();
let {years, months, weeks, days} = duration;
let iter = new Temporal.PlainDate(isoYear + years, isoMonth, isoDay, "iso8601");
const monthsDays = months * 36;
if (iter.dayOfYear + monthsDays > iter.daysInYear || iter.dayOfYear + monthsDays < 1)
  throw new Error("complicated addition not implemented in this test");
return iter.add({ weeks, days: monthsDays + days }).withCalendar(this);
```

guijemont replied:
The code off the top of your head seems to work :)

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/add/custom-daysInMonth-irrelevant.js:
```diff
@@ -0,0 +1,27 @@
+    super("iso8601");
+  }
+  daysInMonth(ym, ...args) {
+      console.log(ym)
+      if (ym.month == 3) {
+          return 15;
+      }
+      return super.daysInMonth(...arguments);
```
To make it simpler, you can just make a completely broken `daysInMonth`:
```suggestion
    return 15;
```

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/subtract/options-undefined.js:
```diff
@@ -15,7 +15,6 @@ class CheckedAdd extends Temporal.Calendar {
   }
   dateAdd(date, duration, options, constructor) {
     this.called = true;
-    assert.notSameValue(options, undefined, "options not undefined");
```
I think we still need to verify this on the 3rd dateAdd call, otherwise the test will no longer be testing what it claims to.

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/subtract/calendar-arguments-extra-options.js:
```diff
@@ -31,7 +31,8 @@ class CustomCalendar extends Temporal.Calendar {
   dateAdd(date, duration, options) {
     const result = super.dateAdd(date, duration, options);
-    options.overflow = 'meatloaf';
+    if (options)
+      options.overflow = 'meatloaf';
```
I didn't realize that this was a side effect of the normative change. Nonetheless, I think it's probably still the correct change to make. It seems to me that the `dateAdd` calls for finding the last day of the month need to have the default options.

You could probably sidestep this by changing the test to subtract -5 months below instead of 5 months.

guijemont replied:
How about adding a `dateAddCalled` counter, and putting a clearer condition? like:
```javascript
    dateAddCalls += 1
    if (dateAddCalls == 3)
      options.overflow = 'meatloaf';
```

guijemont replied:
> How about adding a `dateAddCalled` counter, and putting a clearer condition? like:
> 
> ```js
>     dateAddCalls++;
>     if (dateAddCalls == 3)
>       options.overflow = 'meatloaf';
> ```

I've just pushed a change doing this

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/subtract/calendar-arguments.js:
```diff
@@ -36,7 +36,8 @@ class CustomCalendar extends Temporal.Calendar {
   dateAdd(date, duration, options) {
     const result = super.dateAdd(date, duration, options);
-    options.overflow = 'meatloaf';
+    if (options)
+      options.overflow = 'meatloaf';
```
Same here.

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/subtract/calendar-datefromfields-called.js:
```diff
@@ -77,7 +77,6 @@ class CustomCalendar extends Temporal.Calendar {
     return new Temporal.PlainMonthDay(isoMonth, isoDay, this, isoYear);
   }
   dateAdd(date, duration, options) {
-    if (duration.months) throw new Error("adding months not implemented in this test");
```
Same remark as the corresponding `add` test above.

ptomato commented on test/built-ins/Temporal/PlainYearMonth/prototype/subtract/custom-daysInMonth-irrelevant.js:
```diff
@@ -0,0 +1,27 @@
+    super("iso8601");
+  }
+  daysInMonth(ym, ...args) {
+      console.log(ym)
+      if (ym.month == 3) {
+          return 15;
+      }
+      return super.daysInMonth(...arguments);
```
Same remark as the corresponding `add` test above.

ptomato reviewed (changes requested):
Thanks very much! I have a few suggestions for improvements.

guijemont commented:
> Thanks very much! I have a few suggestions for improvements.

Thanks for the review. I think I addressed all your suggestions, though for the latest commit it's a different approach than the one you suggested.

Let me know if you think I should squash these review commits into the original two commits, I don't know what's the policy on this project. I feel like squashed would be easier for future history review, but I've left them unsquashed for now for ease of review.

ptomato reviewed (approved):
Thanks, all looks good now. Your alternative suggestion was fine.

I do prefer squashing the commits if they form a logical unit together - which is the case with commits that implement review suggestions.

I'll take care of squashing and merging this so that we can then merge the corresponding proposal-temporal PR.
